oga/frontend/public/sw.js

```diff
@@ -1,20 +1,27 @@
 function receivePushNotification(event) {
-  console.log("HI [Service Worker] Push Received.");
-  console.log(event.data.json().text);
+  const { image, tag, url, title, text, data, id, location } = event.data.json();
+  let options = {}
+  switch(tag) {
+    case 'q':
+      options = {
+        data: 'reply/' + id,
+        body: text +' at\n' + location + '?',
+        tag: tag, //this is needed to hide duplicate notifications
+        actions: [{ action: "Detail", title: "Reply" }]
+      }
+    break;
+    case 'a':
+      options = {
+        data: 'reply/' + id,
+        body: text + ' at\n' + location,
+        tag: tag, //this is needed to hide duplicate notifications
+        actions: [{ action: "Detail", title: "Read" }]
```

Comment:
I have question. What is the meaning of 'this is needed to hide duplicate notifications' and actions?

Reply:
Although it is browser dependent, most browsers only show 1 notification with the same tag.
Therefore, if there are 2 notifications about answer, and 1 notification about question, the browser will only show 2, 1 for each tag. Although we do not really need tag data for the user, we can send it to the browser to hide duplicate notifications.